core/src/main/java/com/yahoo/gondola/core/Peer.java

```diff
@@ -163,24 +168,24 @@ public void start() {
         // Start local threads
         assert threads.size() == 0
             : String.format("The threads have not been properly shutdown. %d threads remaining", threads.size());
-        threads.add(new Sender());
-        threads.add(new Receiver());
+        threads.add(new Receiver()); // Close the receiver before sender to avoid "writer is closed" exc from pipedinputstream
         threads.add(new Backfiller());
+        threads.add(new Sender());
         threads.forEach(t -> t.start());
     }
 
     public void stop() {
         generation++;
-        channel.stop();
-        threads.forEach(t -> t.interrupt());
         for (Thread t : threads) {
             try {
+                t.interrupt();
```

Comment:
will it be more effective that interrupt all the threads than join them? it's a serialize operation here.